capa/features/extractors/dnfile/helpers.py

```diff
@@ -299,22 +316,93 @@ def get_dotnet_unmanaged_imports(pe: dnfile.dnPE) -> Iterator[DnUnmanagedMethod]
         # like kernel32.CreateFileA
         yield DnUnmanagedMethod(token, module, method)
 
+def get_nested_class_table(pe):
+    nested_class_table = {}
+    
+    # Used to find nested classes in typedef
+    for rid, nestedclass in iter_dotnet_table(pe, dnfile.mdtable.NestedClass.number):
+        assert isinstance(nestedclass, dnfile.mdtable.NestedClassRow)
+        nested_class_table[nestedclass.NestedClass.row_index] = nestedclass.EnclosingClass.row_index
+
+    return nested_class_table
 
+def get_typedef_class_table(pe):
+    typedef_class_table = []
+    
+    for rid, typedef in iter_dotnet_table(pe, dnfile.mdtable.TypeDef.number):
+        assert isinstance(typedef, dnfile.mdtable.TypeDefRow)
+        typedef_class_table.append((typedef.TypeName, typedef.TypeNamespace))
+        
+    return typedef_class_table
+    
+def typedef_helper(index, nested_class_table, typedef_class_table, n, name):
+    # Append the current typeref name
+    n.append(name)
+    
+    if nested_class_table[index] in nested_class_table:
+        typedef_helper(nested_class_table[index], nested_class_table, typedef_class_table, n, typedef_class_table[nested_class_table[index]-1][0])
+    else:
+        # Document the root enclosing details
+        n.append(typedef_class_table[nested_class_table[index]-1][0])
+        namespace = typedef_class_table[nested_class_table[index]-1][1]
+        n.append(namespace)
+        
+    return tuple(n[::-1])
+
+def get_typeref_table(pe):
+    typeref_table = []
+    
+    # Used to track values in typeref table
+    for rid, typeref in iter_dotnet_table(pe, dnfile.mdtable.TypeRef.number):
+        assert isinstance(typeref, dnfile.mdtable.TypeRefRow)
+        typeref_table.append((typeref.TypeName, typeref.TypeNamespace, typeref.ResolutionScope.row_index, type(typeref.ResolutionScope.table)))
+        
+    return typeref_table
+
+def typeref_helper(index, typeref_table, n, name):
+    # Append the current typeref name
+    n.append(name)
+    
+    if typeref_table[index - 1][3] == dnfile.mdtable.TypeRef:
+        # Recursively call helper function with enclosing typeref details
+        typeref_helper(typeref_table[index - 1][2], typeref_table, n, typeref_table[index - 1][0])
+    else:
+        # Document the root enclosing details
+        n.append(typeref_table[index - 1][0])
+        namespace = typeref_table[index - 1][1]
+        n.append(namespace)
+        
+    return tuple(n[::-1])
+    
 def get_dotnet_types(pe: dnfile.dnPE) -> Iterator[DnType]:
     """get .NET types from TypeDef and TypeRef tables"""
+    nested_class_table = get_nested_class_table(pe)
+    typedef_class_table = get_typedef_class_table(pe)
```

Comment:
Can we instead index the `TypeDef` table directly e.g. `table[rid-1]` or [`table.get_with_row_index(rid)`](https://github.com/malwarefrank/dnfile/blob/b837ad6381a33f85ee0963cfa50f4ad91609b319/src/dnfile/base.py#L875)? I'm suggesting using direct indexing in places like https://github.com/mandiant/capa/blob/94a8b65b480fa0a5944638b0ce3c3dbfd3737bca/capa/features/extractors/dnfile/helpers.py#L345-L348

This avoids iterating the `TypeDef` table more than is needed and makes the indexing more intuitive.

Reply:
Agreed, I changed up the accessing to directly index the `TypeDef` table.